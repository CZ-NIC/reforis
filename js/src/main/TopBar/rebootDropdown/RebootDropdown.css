@keyframes rebootFade {
    from {
        opacity: 0.1;
    }
    to {
        opacity: 1;
    }
}


#reboot-dropdown-toggle {
    animation-name: rebootFade;
    animation-duration: 1.5s;
    animation-iteration-count: infinite;
    animation-direction: alternate;
}

#reboot-dropdown-toggle svg {
    color: red
}
