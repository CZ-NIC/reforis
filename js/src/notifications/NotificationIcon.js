/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import propTypes from 'prop-types';

NotificationIcon.propTypes = {
    severity: propTypes.string.isRequired,
    className: propTypes.string,
};

export default function NotificationIcon({severity, className}) {
    let iconName = null;
    switch (severity) {
        case 'news':
            iconName = 'newspaper';
            break;
        case 'restart':
            iconName = 'power-off';
            break;
        case 'error':
            iconName = 'exclamation-circle';
            break;
        case 'update':
            iconName = 'sync';
            break;
        default:
    }

    return <i className={`fa fa-${iconName} ${className}`}/>;
}