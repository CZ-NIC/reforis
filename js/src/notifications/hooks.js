/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import { useEffect, useState } from "react";

import API_URLs from "common/API";
import { useWSForisModule, useAPIGet, useAPIPost } from "foris";

const WS_MODULE = "router_notifications";

export default function useNotifications(ws) {
    const [notifications, setNotifications] = useState([]);
    const [getState, get] = useAPIGet(API_URLs.notifications);

    useEffect(() => {
        get();
    }, [get]);
    useEffect(() => {
        if (getState.data) {
            const nonDisplayedNotifications = getState.data.notifications
                .filter((notification) => !notification.displayed);
            setNotifications(nonDisplayedNotifications);
        }
    }, [getState]);

    const [WSCreateData] = useWSForisModule(ws, WS_MODULE, "create");
    const [WSMarkAsDisplayedData] = useWSForisModule(ws, WS_MODULE, "mark_as_displayed");
    useEffect(() => {
        if (WSCreateData || WSMarkAsDisplayedData) get();
    }, [WSCreateData, WSMarkAsDisplayedData, get]);

    const [, post] = useAPIPost(API_URLs.notifications);

    function dismiss(notificationId) {
        post({ ids: [notificationId] });
        setNotifications((currentNotifications) => currentNotifications.filter(
            (notification) => notification.id !== notificationId,
        ));
    }

    function dismissAll() {
        post({ ids: notifications.map((notification) => notification.id) });
        setNotifications([]);
    }

    return [notifications, dismiss, dismissAll];
}

const NEW_NOTIFICATION_ANIMATION_DURATION = 1; // Sec.

export function useNewNotification(ws) {
    const [newNotification, setNewNotification] = useState(false);
    useEffect(() => {
        if (newNotification) {
            setTimeout(
                () => setNewNotification(false),
                NEW_NOTIFICATION_ANIMATION_DURATION * 1000,
            );
        }
    }, [newNotification]);

    const [WSCreateData] = useWSForisModule(ws, WS_MODULE, "create");

    useEffect(() => {
        if (WSCreateData) setNewNotification(true);
    }, [WSCreateData]);

    return newNotification;
}
