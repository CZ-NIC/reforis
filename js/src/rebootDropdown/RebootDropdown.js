/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";

import useNotifications from "notifications/hooks";
import RebootButton from "common/RebootButton";
import { ForisURLs, REFORIS_URL_PREFIX } from "foris";
import { Link } from "react-router-dom";

RebootDropdown.propTypes = {
    ws: PropTypes.object.isRequired,
};

export default function RebootDropdown({ ws }) {
    const [notifications] = useNotifications(ws);

    const rebootNotification = notifications.find((notification) => notification.severity === "restart");
    if (!rebootNotification) {
        return null;
    }

    return (
        <div className="dropdown">
            <button type="button" id="reboot-dropdown-toggle" className="nav-item btn btn-link">
                <i className="fa fa-power-off fa-lg" style={{ color: "red" }} />
            </button>
            <div className="dropdown-menu">
                <div className="dropdown-header">
                    <h5>{_("Reboot required")}</h5>
                </div>
                <div className="dropdown-divider" />
                <div className="dropdown-item">
                    <Link to={{ pathname: ForisURLs.notifications, search: `?id=${rebootNotification.id}` }}>
                        <button type="button" className="btn btn-primary mr-3">{_("Details")}</button>
                    </Link>
                    <RebootButton />
                </div>
            </div>
        </div>
    );
}
