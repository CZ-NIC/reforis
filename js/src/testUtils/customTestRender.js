/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

/* eslint import/export: "off" */

import React from 'react';
import {UIDReset} from 'react-uid';
import {StaticRouter} from 'react-router';
import {render} from '@testing-library/react'

function Wrapper({children}) {
    return <StaticRouter>
        <UIDReset>
            {children}
        </UIDReset>
    </StaticRouter>
}

const customTestRender = (ui, options) => render(ui, {wrapper: Wrapper, ...options});

// re-export everything
export * from '@testing-library/react'

// override render method
export {customTestRender as render}
