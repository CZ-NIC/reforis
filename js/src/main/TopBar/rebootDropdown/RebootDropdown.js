/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";

import PropTypes from "prop-types";
import { ForisURLs } from "foris";
import { Link } from "react-router-dom";

import useNotifications from "notifications/hooks";
import RebootButton from "common/RebootButton";
import smallScreenWidth from "../../../utils/constants";

import "./RebootDropdown.css";

RebootDropdown.propTypes = {
    ws: PropTypes.object.isRequired,
};

export default function RebootDropdown({ ws }) {
    const [notifications] = useNotifications(ws);

    const rebootNotification = notifications.find(
        (notification) => notification.severity === "restart"
    );
    if (!rebootNotification) {
        return null;
    }

    return (
        <div className="dropdown">
            <button
                type="button"
                id="reboot-dropdown-toggle"
                className="nav-item btn btn-link"
            >
                <i className="fas fa-power-off fa-lg" />
            </button>
            <div
                className={`dropdown-menu dropdown-menu-${
                    window.outerWidth > smallScreenWidth ? "right" : "left"
                } shadow-sm`}
            >
                <div className="dropdown-header">
                    <Link
                        to={{
                            pathname: ForisURLs.rebootPage,
                        }}
                    >
                        <h5>{_("Reboot Required")}</h5>
                    </Link>
                </div>
                <div className="dropdown-divider" />
                <div className="dropdown-item">
                    <Link
                        to={{
                            pathname: ForisURLs.overview,
                            search: `?id=${rebootNotification.id}`,
                        }}
                    >
                        <button type="button" className="btn btn-primary mr-3">
                            {_("Details")}
                        </button>
                    </Link>
                    <RebootButton />
                </div>
            </div>
        </div>
    );
}
