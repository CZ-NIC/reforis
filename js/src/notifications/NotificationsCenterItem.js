/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */
import React from 'react';
import NotificationIcon from "./NotificationIcon";
import {toLocaleDateString} from "./utils";

const BORDER_TYPES = {
    news: 'border-info',
    update: 'border-info',
    restart: 'border-danger',
    error: 'border-danger',
};

export default class NotificationsCenterItem extends React.Component {
    constructor(props) {
        super(props);
        this.myRef = React.createRef();
    }

    componentDidMount() {
        // Hack to scroll to needed item. ForisNotificationID is defined in the notification center template.
        if (window.ForisNotificationID === undefined)
            return;
        if (window.ForisNotificationID === this.props.notification.id)
            this.myRef.current.scrollIntoView({block: 'start', behavior: 'smooth'});
    }

    render() {
        const notification = this.props.notification;
        return <div ref={this.myRef} className={`card bg-light ${BORDER_TYPES[notification.severity]} sm-10`}>
            <div className="card-header">
                <NotificationIcon severity={notification.severity} className={'fa-2x'}/>
                <p className="text-muted">{toLocaleDateString(notification.created_at)}</p>
                <button
                    type="button"
                    className="close"
                    onClick={e => this.props.dismissHandler(notification.id)}
                >Ã—
                </button>
            </div>

            <div className="card-body">
                <p className="card-text">{notification.msg}</p>
            </div>
        </div>;
    }
}

