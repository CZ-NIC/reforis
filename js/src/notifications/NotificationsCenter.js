/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */


import React from "react";
import NotificationsWrapper from "./NotificationsWrapper";
import NotificationsCenterItem from "./NotificationsCenterItem";

class NotificationsCenter extends React.PureComponent {
    getDismissAllButton() {
        return <button
            type="button"
            id="btn-dismiss-all"
            className="btn btn-outline-danger float-right"
            onClick={this.props.dismissAllHandler}
        >
            {_('Dismiss all')}
        </button>
    }

    render() {

        return <div id='notifications-center' className="col-sm-10">
            <h3>{_('Settings')}</h3>
            <h3>{_('Notifications')}</h3>
            {
                this.props.notifications.length !== 0 ?
                    this.getDismissAllButton() :
                    <p className="text-muted text-center">{_('No notifications')}</p>
            }

            <NotificationsCenterItemsList
                notifications={this.props.notifications}

                dismissHandler={this.props.dismissHandler}
            />
        </div>
    }
}

function NotificationsCenterItemsList({notifications, dismissHandler}) {
    return notifications.map(
        notification => {
            return <NotificationsCenterItem
                key={notification.id}
                notification={notification}
                dismissHandler={dismissHandler}
            />
        }
    )
}

export default NotificationsWrapper(NotificationsCenter);