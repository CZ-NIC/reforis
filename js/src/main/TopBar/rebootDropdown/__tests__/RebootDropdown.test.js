/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { render, wait } from "foris/testUtils/customTestRender";

import { notificationsFixture } from "../../../../notifications/__tests__/__fixtures__/notifications";

import RebootDropdown from "../RebootDropdown";

describe("<RebootDropdown/>", () => {
    let rebootDropdownContainer;

    beforeEach(async () => {
        const { container, getByText } = render(
            <RebootDropdown
                notifications={notificationsFixture.notifications}
            />
        );
        rebootDropdownContainer = container;
        await wait(() => {
            getByText("Reboot Required");
        });
    });

    it("Test with snapshot", () => {
        expect(rebootDropdownContainer).toMatchSnapshot();
    });
});
