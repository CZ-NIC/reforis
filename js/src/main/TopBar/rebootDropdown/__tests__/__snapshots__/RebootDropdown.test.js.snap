// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RebootDropdown/> Test with snapshot 1`] = `
<div>
  <div
    class="dropdown"
  >
    <button
      class="nav-item btn btn-link"
      id="reboot-dropdown-toggle"
      type="button"
    >
      <i
        class="fas fa-power-off fa-lg"
      />
    </button>
    <div
      class="dropdown-menu dropdown-menu-right shadow-sm"
    >
      <div
        class="dropdown-header"
      >
        <a
          href="/reforis/administration/reboot"
        >
          <h5>
            Reboot Required
          </h5>
        </a>
      </div>
      <div
        class="dropdown-divider"
      />
      <div
        class="dropdown-item"
      >
        <a
          href="/overview?id=808-909"
        >
          <button
            class="btn btn-primary mr-3"
            type="button"
          >
            Details
          </button>
        </a>
        <button
          class="btn btn-danger"
          type="button"
        >
          Reboot
        </button>
      </div>
    </div>
  </div>
</div>
`;
